checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -43,104 +92,621 @@ public class MustCallInferenceLogic {
+    enclosingMethodElt = TreeUtils.elementFromDeclaration(enclosingMethodTree);
+  }
+
+  /**
+   * Creates a MustCallInferenceLogic instance, runs the inference algorithm. If the type factory
+   * has whole program inference enabled, its postAnalyze method should execute the inference
+   * algorithm using this method.
+   *
+   * @param typeFactory the type factory
+   * @param cfg the control flow graph of the method to check
+   * @param mcca the MustCallConsistencyAnalyzer
+   */
+  protected static void runMustCallInferenceLogic(
+      ResourceLeakAnnotatedTypeFactory typeFactory,
+      ControlFlowGraph cfg,
+      MustCallConsistencyAnalyzer mcca) {
+    MustCallInferenceLogic mustCallInferenceLogic =
+        new MustCallInferenceLogic(typeFactory, cfg, mcca);
+    mustCallInferenceLogic.runInference();
   }
 
   /**
-   * Runs the inference algorithm on the contents of the {@link #cfg} field.
+   * Runs the inference algorithm on the contents of the {@link #cfg} field and parameters.
    *
-   * <p>Operationally, it checks method invocations for fields with non-empty @MustCall obligations
-   * along all paths to the regular exit point in the method body of the method represented by
-   * {@link #cfg}, and updates the {@link #owningFields} set if it discovers an owning field whose
-   * must-call obligations were satisfied along one of the checked paths.
+   * <p>Operationally, it checks method invocations for fields and parameters with
+   * non-empty @MustCall obligations along all paths to the regular exit point in the method body of
+   * the method represented by {@link #cfg}, and updates the {@link #owningFieldToECM} set or
+   * adds @Owning to the formal parameter if it discovers their must-call obligations were satisfied
+   * along one of the checked paths.
    */
-  /*package-private*/ void runInference() {
-    Set<Block> visited = new HashSet<>();
-    Deque<Block> worklist = new ArrayDeque<>();
-    Block entry = this.cfg.getEntryBlock();
+  private void runInference() {
+
+    Set<BlockWithObligations> visited = new HashSet<>();
+
+    Deque<BlockWithObligations> worklist = new ArrayDeque<>();
+
+    BlockWithObligations entry =
+        new BlockWithObligations(cfg.getEntryBlock(), getNonEmptyMCParams(cfg));
 
     worklist.add(entry);
     visited.add(entry);
 
     while (!worklist.isEmpty()) {
-      Block current = worklist.remove();
+      BlockWithObligations current = worklist.remove();
 
-      for (Node node : current.getNodes()) {
-        if (node instanceof MethodInvocationNode) {
-          checkForMustCallInvocationOnField((MethodInvocationNode) node);
+      Set<Obligation> obligations = new LinkedHashSet<>(current.obligations);
+
+      for (Node node : current.block.getNodes()) {
+        if (node instanceof MethodInvocationNode || node instanceof ObjectCreationNode) {
+          mcca.updateObligationsWithInvocationResult(obligations, node);
```

Comment:
This call will not cause any side effects, since the inference takes place within the postAnalyze method of the ResourceLeakAnnotatedTypeFactory, then the consistency analyzer is finished.